File: modules/io/examples/mesh_convert.cpp
/*
 * Copyright 2023 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
#include <lagrange/Logger.h>
#include <lagrange/io/load_mesh.h>
#include <lagrange/io/load_simple_scene.h>
#include <lagrange/io/save_mesh.h>
#include <lagrange/io/save_simple_scene.h>
#include <lagrange/mesh_cleanup/remove_duplicate_vertices.h>
#include <lagrange/scene/simple_scene_convert.h>
#include <lagrange/utils/strings.h>

#include <CLI/CLI.hpp>

#include <unordered_set>

using Index = uint32_t;
constexpr size_t Dimension = 3;

std::unordered_set<std::string_view> input_scene_formats()
{
    return {".gltf", ".glb", ".fbx", ".lgs"};
}
std::unordered_set<std::string_view> output_scene_formats()
{
    return {".gltf", ".glb", ".obj", ".lgs"};
}

template <typename Scalar>
void convert(const lagrange::fs::path& input_filename, const lagrange::fs::path& output_filename)
{
    using MeshType = lagrange::SurfaceMesh<Scalar, Index>;
    using SceneType = lagrange::scene::SimpleScene<Scalar, Index, Dimension>;

    // Load as scene if extension is .fbx, .gtlf or .glb
    std::string input_ext = lagrange::to_lower(input_filename.extension().string());
    if (input_scene_formats().count(input_ext)) {
        // Load scene
        lagrange::logger().info("Loading input scene: {}", input_filename.string());
        lagrange::io::LoadOptions load_options;
        if (input_ext == ".gltf" || input_ext == ".glb") {
            // If input scene if gtTF, we need to stitch duplicate vertices (glTF doesn't support
            // indexed buffers)
            load_options.stitch_vertices = true;
        }
        auto scene = lagrange::io::load_simple_scene<SceneType>(input_filename, load_options);

        // Display info
        lagrange::logger().info(
            "Input scene has {} meshes and {} instances",
            scene.get_num_meshes(),
            scene.compute_num_instances());

        // Save as scene or mesh
        std::string output_ext = lagrange::to_lower(output_filename.extension().string());
        if (output_scene_formats().count(output_ext)) {
            lagrange::logger().info("Saving output scene: {}", output_filename.string());
            lagrange::io::save_simple_scene(output_filename, scene);
        } else {
            lagrange::logger().info("Saving output mesh: {}", output_filename.string());
            lagrange::TransformOptions options;
            options.reorient = true;
            lagrange::io::save_mesh(
                output_filename,
                lagrange::scene::simple_scene_to_mesh(scene, options));
        }
    } else {
        // Load mesh
        lagrange::logger().info("Loading input mesh: {}", input_filename.string());
        auto mesh = lagrange::io::load_mesh<MeshType>(input_filename);

        // Display info
        lagrange::logger().info(
            "Input mesh has {} vertices and {} faces",
            mesh.get_num_vertices(),
            mesh.get_num_facets());

        // Save mesh
        lagrange::logger().info("Saving output mesh: {}", output_filename.string());
        lagrange::io::save_mesh(output_filename, mesh);
    }
}

int main(int argc, char** argv)
{
    struct
    {
        std::string input;
        std::string output;
        bool use_float = true;
    } args;

    CLI::App app{argv[0]};
    app.option_defaults()->always_capture_default();
    app.add_option("input", args.input, "Input mesh.")->required()->check(CLI::ExistingFile);
    app.add_option("output", args.output, "Output mesh.")->required();
    app.add_option("--float", args.use_float, "Use float precision.");
    CLI11_PARSE(app, argc, argv)

    if (args.use_float) {
        convert<float>(args.input, args.output);
    } else {
        convert<double>(args.input, args.output);
    }

    return 0;
}

File: modules/core/python/scripts/meshconvert.py
#!/usr/bin/env python

#
# Copyright 2024 Adobe. All rights reserved.
# This file is licensed to you under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License. You may obtain a copy
# of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under
# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
# OF ANY KIND, either express or implied. See the License for the specific language
# governing permissions and limitations under the License.
#
import argparse
import lagrange
from pathlib import Path


def parse_args():
    parser = argparse.ArgumentParser(description="Convert a mesh file to a different format.")
    parser.add_argument("input_mesh", help="input mesh file")
    parser.add_argument("output_mesh", help="output mesh file")
    parser.add_argument("--triangulate", "-t", action="store_true", help="triangulate the mesh")
    parser.add_argument("--stitch-vertices", '-s', action="store_true", help="stitch vertices")
    parser.add_argument(
        "--logging-level",
        "-l",
        default="info",
        help="logging level",
        choices=["debug", "info", "warning", "error", "critical"],
    )
    return parser.parse_args()


def main():
    args = parse_args()
    lagrange.logger.setLevel(args.logging_level.upper())
    input_suffix = Path(args.input_mesh).suffix
    output_suffix = Path(args.output_mesh).suffix
    if input_suffix in [".fbx", ".gltf", ".glb"]:
        lagrange.logger.info(f"Loading input scene from {args.input_mesh}")
        scene = lagrange.io.load_simple_scene(args.input_mesh)
        if input_suffix in [".gltf", ".glb"]:
            for i in range(scene.num_meshes):
                mesh = scene.ref_mesh(i)
                lagrange.remove_duplicate_vertices(mesh)

        if args.triangulate:
            for i in range(scene.num_meshes):
                mesh = scene.ref_mesh(i)
                lagrange.triangulate_polygonal_facets(mesh)

        if output_suffix in [".gltf", ".glb"]:
            lagrange.logger.info(f"Saving output scene in {args.output_mesh}")
            lagrange.io.save_simple_scene(args.output_mesh, scene)
        else:
            lagrange.logger.info(f"Saving output mesh in {args.output_mesh}")
            mesh = lagrange.scene.simple_scene_to_mesh(scene)
            lagrange.io.save_mesh(args.output_mesh, mesh)
    else:
        lagrange.logger.info(f"Loading input mesh from {args.input_mesh}")
        mesh = lagrange.io.load_mesh(args.input_mesh, stitch_vertices=args.stitch_vertices)
        if not mesh.is_regular and args.triangulate:
            lagrange.triangulate_polygonal_facets(mesh)
        lagrange.logger.info(f"Saving output mesh in {args.output_mesh}")
        lagrange.io.save_mesh(args.output_mesh, mesh)


if __name__ == "__main__":
    main()
